proof/refine/ARM/Finalise_R.thy

```diff
@@ -2326,6 +2326,342 @@ lemma replyClear_invs'[wp]:
   apply wpsimp
   sorry
 
+lemma isHeadSome:
+ "isHead (Some h) \<longleftrightarrow> (\<exists>p. h = Head p)"
+  by (clarsimp simp: isHead_def split: reply_next.split)
+
+lemma in_list_refs_of_replies'_E:
+  assumes in_list: "(p, tp) \<in> list_refs_of_replies' s replyPtr"
+  assumes nextR: "\<And>reply. tp = ReplyNext \<Longrightarrow> replyNext reply = Some (Next p)
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  assumes prevR: "\<And>reply. tp = ReplyPrev \<Longrightarrow> replyPrev reply = Some p
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  shows "R"
+  using in_list
+  apply (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def in_get_refs
+                 split: option.splits)
+  by (auto elim: nextR prevR simp: opt_map_def)
+
+lemma updateReply_ko_at'_reply_sp:
+  "\<lbrace>ko_at' reply replyPtr\<rbrace>
+   updateReply replyPtr (replyTCB_update Map.empty)
+   \<lbrace>\<lambda>_. ko_at' (reply\<lparr>replyTCB := None\<rparr>) replyPtr\<rbrace>"
+  apply (simp add: updateReply_def)
+  apply (wpsimp wp: set_reply'.set_ko_at')
+  using ko_at'_inj by blast
+
+lemma replyUnlink_ko_at'_reply_sp:
```

Comment:
Are these sp rules (this one and the above) used somewhere? If not, probably we don’t need to keep them.

Reply:
Removed